[v3] Implement advanced shard awareness

Lorak-mmk opened the pull request:
This is continuation of #36. I addresed comments by @avelanarius - https://github.com/scylladb/java-driver/pull/36#issuecomment-903680419.

TODO:

- [x] When closing connection, free port in PortAllocator
- [x] In PortAllocator check if port is really free before returning it - I expect it should be faster (when many ports are busy) than opening Netty's channel multiple times.
- [x] Manual testing
- [x] There may be some unnecessary debug prints left, remove them.
- [x] When trying next local port, close previous channel.
- [x] Refactor initAsync changes - they are a bit ugly now.

Commit:
Limit the rate at which HostConnectionPool opens connections

Changed test:
New reconnection mechanism caused one test to fail.
Test needed to be changed, because missing connections will now be
spawned by the reconnection mechanism - no need to send requests
to schedule connection creation.

Commit:
Fix borrowConnection handling of missing connection to a shard

Previously, if there was no connection to a shard and
host.convictionPolicy.canReconnectNow() was true, method would wait for
the new connection to be established. This is not correct because we would like
to use connection to a different shard if available.

This change fixes the problem.

Commit:
Make it possible to create connection using different server port

This will be used to establish connection to the shard-aware port.

Commit:
Fetch server ports with advanced shard awareness support

Driver needs to fetch shard-aware port on first connection to node,
so that subsequent connection can be shard aware.

There is a separate shard-aware port for SSL connections,
just as there is one for non shard-aware connections.

ShardingInfo needs to expose the right port, depending on whether SSL is
enabled or not.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/Connection.java:
```diff
@@ -106,6 +111,10 @@
   private static final int FLUSHER_SCHEDULE_PERIOD_NS =
       SystemProperties.getInt("com.datastax.driver.FLUSHER_SCHEDULE_PERIOD_NS", 10000);
 
+  private static final long ADV_SHARED_AWARENESS_BLOCK_ON_NAT = 1000000L * 60L * 1000L;
```
```suggestion
  private static final long ADV_SHARD_AWARENESS_BLOCK_ON_NAT = 1000000L * 60L * 1000L;
```

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/Connection.java:
```diff
@@ -106,6 +111,10 @@
 
+  private static final long ADV_SHARED_AWARENESS_BLOCK_ON_NAT = 1000000L * 60L * 1000L;
+
+  private static final long ADV_SHARED_AWARENESS_BLOCK_ON_ERROR = 5 * 60 * 1000;
```
```suggestion
  private static final long ADV_SHARD_AWARENESS_BLOCK_ON_ERROR = 5 * 60 * 1000;
```

Lorak-mmk commented:
The only thing left are the manual tests - I think the PR can be reviewed now.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/Connection.java:
```diff
@@ -1039,6 +1159,60 @@ public String toString() {
         "Connection[%s, inFlight=%d, closed=%b]", name, inFlight.get(), isClosed());
   }
 
+  static class PortAllocator {
```
Isn't it a problem if `startPort` eventually gets high (near `highPort`) and `getNextAvailablePort` won't be able to find a free port? In that case it should probably wrap and start again from `lowPort`.

Also, isn't `busyPorts` array a bit of an overkill? The solution I had in mind was to have a single "global" `lastPort` and `getNextAvailablePort` would bump it higher and higher. By only increasing it, the risk of collisions would be smaller.

I had something like this in mind:

```java
// static for the sake of example
private static final AtomicInteger lastPort = new AtomicInteger(-1);

public static int getNextAvailablePort(int shardCount, int shardId, int lowPort, int highPort) {
    int lastPortValue, foundPort = -1;
    do {
        lastPortValue = lastPort.get();

        // We will scan from lastPortValue
        // (or lowPort is there was no lastPort)
        int scanStart = lastPortValue == -1 ? lowPort : lastPortValue;

        // Round it up to "% shardCount == shardId"
        scanStart += (shardCount - scanStart % shardCount) + shardId;
        assert scanStart % shardCount == shardId;

        // Scan from scanStart upwards to highPort.
        for (int port = scanStart; port <= highPort; port += shardCount) {
            if (isTcpPortAvailable(port)) {
                foundPort = port;
                break;
            }
        }

        // If we started scanning from a high scanStart port
        // there might have been not enough ports left that are
        // smaller than highPort. Scan from the beginning
        // from the lowPort.
        if (foundPort == -1) {
            scanStart = lowPort + (shardCount - lowPort % shardCount) + shardId;
            assert scanStart % shardCount == shardId;

            for (int port = scanStart; port <= highPort; port += shardCount) {
                if (isTcpPortAvailable(port)) {
                    foundPort = port;
                    break;
                }
            }
        }

        // No luck! All ports taken!
        if (foundPort == -1) {
            return -1;
        }
    } while (!lastPort.compareAndSet(lastPortValue, foundPort));

    assert foundPort % shardCount == shardId
            && foundPort >= lowPort
            && foundPort <= highPort;
    return foundPort;
}
```

Lorak-mmk replied:
> Isn't it a problem if startPort eventually gets high (near highPort) and getNextAvailablePort won't be able to find a free port? In that case it should probably wrap and start again from lowPort.

For each Connection, `[lowPort, highPort]` range will be search exactly once.

> Also, isn't busyPorts array a bit of an overkill? The solution I had in mind was to have a single "global" lastPort and getNextAvailablePort would bump it higher and higher. By only increasing it, the risk of collisions would be smaller.

Your solution is definitely better in a good case (low % of busy ports), but in a bad case (high % of busy ports), won't it be worse to try to open socket on each port rather than just search an array? Anyway, good case is probably more important, I'll apply your solution (modified a bit, as it needs to take into consideration that there may be multiple sessions).

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -1473,6 +1476,36 @@ public Builder withCloudSecureConnectBundle(InputStream cloudConfigInputStream)
+     * was called, then setting this range does not affect anything.
+     *
+     * @param low Lower bound of range, inclusive.
+     * @param high Upper bound of range, inclusive.
```
Missing `@return this builder`. If I remember correctly, this will make JavaDoc generation fail and JavaDoc generation is a necessary step when publishing a package to Maven.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -1473,6 +1476,36 @@ public Builder withCloudSecureConnectBundle(InputStream cloudConfigInputStream)
+     * Disables advanced shard awareness. By default, this driver chooses local port while making a
+     * connection to node, to signal which shard it wants to connect to. This allows driver to
+     * estabilish connection pool faster, especially when there are multiple clients connecting
+     * concurrently. If this causes any issues, you can disable it using this method.
```
Nit: "If this causes any issues" - describe that a NAT or network misconfiguration (i.e. shard-aware port 19042 not exposed to the external world) is a common cause.

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -1473,6 +1476,36 @@ public Builder withCloudSecureConnectBundle(InputStream cloudConfigInputStream)
+     * @param high Upper bound of range, inclusive.
+     */
+    public Builder withLocalPortRange(int low, int high) {
+      if (low < 1 || 65535 < low || 1 < high || 65535 < high) {
```
Nit: `low < high` and also some requirement that this range is sufficiently large (`low = 10000, high = 10001` not accepted).

Lorak-mmk replied:
What minimum amount of ports in range would you propose? 1000?

Lorak-mmk commented:
All test scenarios from https://github.com/scylladb/java-driver/pull/36#issuecomment-903680419 are passing with current version of PR.

piodul commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -1473,6 +1476,44 @@ public Builder withCloudSecureConnectBundle(InputStream cloudConfigInputStream)
+     * @return this builder.
+     */
+    public Builder withLocalPortRange(int low, int high) {
+      if (low < 1 || 65535 < low || 1 < high || 65535 < high) {
```
`1 < high`

Shouldn't this be `high < 1`?

haaawk commented:
I think this PR needs some rebasing no? It seems that commits are just a log of work over time not the cleaned up version of the change. Or do I have a wrong impression?

avelanarius commented:
Tested this PR on top of my recent work in #94. Unfortunately, this PR caused regression in 4 integration tests: https://github.com/avelanarius/java-driver/pull/2/checks?check_run_id=3547107895

At first glance it looks like those failures are caused by changed parameters of `Connection.Factory::open()` (and mock expecting a different call), so should be easy to fix. Please refer to #94 for instructions on how to start integration tests ~~(maybe by tomorrow I will have that PR merged)~~ (when you rebase this PR on top of latest scylla-3.x branch, the tests will automatically start on GitHub Actions).

Lorak-mmk committed:
Make advanced shard awareness configurable

Create public API to configure port-based shard awareness.
User can disable it with Cluster.Builder.withoutAdvancedShardAwareness()
and set range of local ports used with
Cluster.Builder.withLocalPortRange(int low, int high).

Lorak-mmk committed:
Implement PortAllocator

PortAllocator finds a free port for a given shard in a given range.
It will be used by port-based shard awareness.
Important think to remember is that returned port may be stolen before
we can use it, so to correctly bind to a local port we need to,
in a loop, get a free port and try to bind it.

Lorak-mmk committed:
Implement fallback to non-advanced shard awareness

When we connect to different shard than requested, then probably NAT is
present - we should disable advanced shard awareness and use old
strategy.

When we failed to connect to node using shard-aware port, then we
disable advanced shard awareness for 5 minutes - hopefully connection
problems will be fixed by then.

Lorak-mmk committed:
Use port-based shard awareness

Use advanced (port-based) shard awareness both when creating initial
connection pool, and later when creating missing connections.

Lorak-mmk committed:
Make it possible to connect to given shard in Connection

Lorak-mmk committed:
Retry connection when local port is busy

Local port may be stolen before we open channel, in that case we need to
try again.

Lorak-mmk commented:
Created better commit history, rebased on current version of branch, fixed some of failing tests (I'll fix the rest soon).

Lorak-mmk commented:
I haven't yet been able to debug the last failing test, it's more problematic than the others.

haaawk commented on driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:
```diff
@@ -421,14 +421,12 @@ private Connection findLeastBusyForShard(int shardId) {
       if (host.convictionPolicy.canReconnectNow()) {
         if (connectionsPerShard == 0) {
```
It seems that this case when connectionsPerShard == 0 could be removed at all. I can't think of a case when it's true. @avelanarius @Lorak-mmk please verify.

haaawk replied:
On a second thought it might be guarding against a configuration where core connections == 0 so it's better to keep it.

avelanarius commented:
@Lorak-mmk regarding this one integration test that fails, the problem is as follows:

In `addConnectionIfUnderMaximum` if `convictionPolicy` says that the host cannot be reconnected to, it returns `false`:

https://github.com/scylladb/java-driver/blob/4fce0af640b4f0f7be54b67217236b38e064a6fc/driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java#L743-L746

This return value goes into `ConnectionTask::run()`:

https://github.com/scylladb/java-driver/blob/4fce0af640b4f0f7be54b67217236b38e064a6fc/driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java#L211-L218

which schedules another reconnection try. There are other `return false;` in `addConnectionIfUnderMaximum` which are now retried (ad infinitum!) under new `ConnectionTask::run()`.

I will try to think of a way to fix this problem or how to implement connection establishment throttling another way (which does not introduce this new "retry" behavior).

avelanarius commented on driver-core/src/test/java/com/datastax/driver/core/HostConnectionPoolTest.java:
```diff
@@ -1162,7 +1163,7 @@ public void should_not_schedule_reconnect_when_connection_lost_and_at_core_conne
   @Test(groups = "short")
   public void should_not_mark_host_down_if_some_connections_fail_on_init() throws Exception {
-    int readTimeout = 1000;
+    int readTimeout = 000;
```
This seems like a mistake?

Lorak-mmk replied:
Fixed

avelanarius commented on driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:
```diff
@@ -671,22 +753,36 @@ private boolean addConnectionIfUnderMaximum(int shardId, ConnectionTasksSharedSt
         if (!host.convictionPolicy.canReconnectNow()) {
           open[shardId].decrementAndGet();
-          return false;
+          return ConnectionResult.SHOULD_RETRY;
```
Why not make it `ConnectionResult.FAILED`, preserving the old behavior?

haaawk replied:
New behavior is just better.

avelanarius commented on driver-core/src/test/java/com/datastax/driver/core/HostConnectionPoolTest.java:
```diff
@@ -1197,16 +1198,11 @@ public void should_not_mark_host_down_if_some_connections_fail_on_init() throws
-        verify(factory, timeout(readTimeout)).open(any(HostConnectionPool.class));
-        reset(factory);
-        assertPoolSize(pool, i);
-      }
+      // Reconnection mechanism should fill missing connections by now.
+      verify(factory, timeout(readTimeout).times(4))
+          .open(any(HostConnectionPool.class), anyInt(), anyInt());
+      reset(factory);
+      assertPoolSize(pool, 8);
```
I think that the old code can be made working - by making the change in my previous comment (related to `ConnectionResult.FAILED` in one case).

haaawk replied:
Isn't the new behavior nicer?

haaawk reviewed (approved):
LGTM @avelanarius do you think we're good to merge this PR?

CatTail commented:
Hi @haaawk @avelanarius, it seems the maven repo missing the latest improvement you made in java driver, can we publish a new version including these changes?

I also have a question regarding how to know whether the application is using new advanced shard aware or the old one, as far as I know, we will fallback to the old shard aware client driver can't connect to advanced shard aware port, but is there any logs or monitor available to make sure which one is actually being used?

Thank you!

haaawk commented:
We will try to publish a new version to maven this week, @CatTail. Thanks for mentioning this.

haaawk commented:
At the moment we don't log whether advanced shard awarness is used or not but we will add the log before releasing. I created https://github.com/scylladb/java-driver/issues/112. Thanks for a very good the suggestion @CatTail!

CatTail commented:
Thank you @haaawk!

CatTail commented:
Hi @haaawk, can I ask do we have any progress on publishing the new version scylla driver?

haaawk commented:
@avelanarius ?

avelanarius commented:
@CatTail Scylla Java Driver `3.11.0.0` is now released and contains this change (note that to be consistent with 4.x branch we changed the naming from `3.X.Y-scylla-Z` to `3.X.Y.Z`).

CatTail commented:
Hi @avelanarius thanks for helping on this, not sure if I look at the right place, but the version `3.11.0.0` doesn't exist in https://mvnrepository.com/artifact/com.scylladb/scylla-driver-core ?

avelanarius commented:
@CatTail It needs some time to fully propagate to all those websites. However, it is live in Maven Central Repository: https://repo1.maven.org/maven2/com/scylladb/scylla-driver-core/3.11.0.0/ (although if it is behind some CDN, it might not be visible yet - nothing we can do about it apart from waiting a few hours).

CatTail commented:
Thank you @avelanarius, Yes I can confirm the artifact works as expected.
